pkg/genericapiserver/api/handlers/rest.go

```diff
@@ -570,25 +572,30 @@ func patchResource(
 		}
 
 		switch {
-		case len(originalObjJS) == 0 || len(originalPatchedObjJS) == 0:
+		case originalObjJS == nil && originalObjMap == nil:
 			// first time through,
 			// 1. apply the patch
-			// 2. save the originalJS and patchedJS to detect whether there were conflicting changes on retries
-			if js, err := runtime.Encode(codec, currentObject); err != nil {
-				return nil, err
-			} else {
-				originalObjJS = js
-			}
-
-			if js, err := getPatchedJS(patchType, originalObjJS, patchJS, versionedObj); err != nil {
-				return nil, err
-			} else {
-				originalPatchedObjJS = js
-			}
+			// 2. save the original and patched to detect whether there were conflicting changes on retries
 
 			objToUpdate := patcher.New()
-			if err := runtime.DecodeInto(codec, originalPatchedObjJS, objToUpdate); err != nil {
-				return nil, err
+
+			// For performance reasons, in case of strategicpatch, we avoid json
+			// marshaling and unmarshaling and operate just on map[string]interface{}.
+			// In case of other patch types, we still have to operate on JSON
+			// representations.
+			switch patchType {
+			case types.JSONPatchType, types.MergePatchType:
+				if originalJS, patchedJS, err := patchObjectJSON(patchType, codec, currentObject, patchJS, objToUpdate, versionedObj); err != nil {
+					return nil, err
+				} else {
```

Comment:
Remove else's on returns, use assign outside the if (the if's are really too long anyway)